README.md
# stactools-msbuildings

[![PyPI](https://img.shields.io/pypi/v/stactools-msbuildings)](https://pypi.org/project/stactools-msbuildings/)

- Name: msbuildings
- Package: `stactools.msbuildings`
- PyPI: https://pypi.org/project/stactools-msbuildings/
- Owner: @TomAugspurger
- Dataset homepage: http://example.com
- STAC extensions used:
  - [table](https://github.com/stac-extensions/table/)
- Extra fields:
  - `msbuildings:region`: The region covered by an item.

This package generates STAC items for the Bing Maps building footprints dataset.

## STAC Examples

- [Collection](examples/collection.json)
- [Item](examples/item/item.json)

## Installation
```shell
pip install stactools-msbuildings
```

## Command-line Usage

Description of the command line functions

```shell
$ stac msbuildings create-item source destination
```

Use `stac msbuildings --help` to see all subcommands and options.

## Contributing

We use [pre-commit](https://pre-commit.com/) to check any changes.
To set up your development environment:

```shell
$ pip install -e .
$ pip install -r requirements-dev.txt
$ pre-commit install
```

To check all files:

```shell
$ pre-commit run --all-files
```

To run the tests:

```shell
$ pytest -vv
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [0.2.0]

### Changed

- Changed the item ID format to include a "quadkey", in addition to the processing date and region.

### Added

- Added a new `msbuildings:quadkey` property.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

[Unreleased]: <https://github.com/stactools-packages/msbuildings/tree/main/>


tests/test_stac.py
import requests

from stactools.msbuildings import stac


def test_collection() -> None:
    result = stac.create_collection()
    assert result.links
    assert result.keywords
    assert "thumbnail" in result.assets
    assert "item_assets" in result.extra_fields


def test_create_item() -> None:
    token = requests.get(
        "https://planetarycomputer-staging.microsoft.com/api/sas/v1/token/bingmlbuildings/footprints"  # noqa: E501
    ).json()["token"]
    asset_href = "abfs://footprints/delta/2023-04-25/ml-buildings.parquet/RegionName=Abyei/quadkey=122321003/"  # noqa: E501

    item = stac.create_item(
        asset_href,
        asset_extra_fields={
            "table:storage_options": {"account_name": "bingmlbuildings"}
        },
        storage_options={"account_name": "bingmlbuildings", "credential": token},
    )
    item.validate()

    assert item.id == "Abyei_122321003_2023-04-25"
    assert item.bbox
    assert item.geometry
    assert item.datetime is None
    assert item.properties["start_datetime"] == "2016-03-31T00:00:00+00:00"
    assert item.properties["end_datetime"] == "2016-06-20T00:00:00+00:00"
    assert item.properties["msbuildings:region"] == "Abyei"
    assert item.properties["msbuildings:quadkey"] == 122321003
    assert item.properties["msbuildings:processing-date"] == "2023-04-25"
    assert item.assets["data"].to_dict() == {
        "href": "abfs://footprints/delta/2023-04-25/ml-buildings.parquet/RegionName=Abyei/quadkey=122321003/",  # noqa: E501
        "type": "application/x-parquet",
        "title": "Building Footprints",
        "description": "Parquet dataset with the building footprints for this region.",
        "table:storage_options": {"account_name": "bingmlbuildings"},
        "roles": ["data"],
    }

    assert item.properties["table:columns"] == [
        {
            "description": "Building footprint polygons",
            "name": "geometry",
            "type": "byte_array",
        },
        {"name": "meanHeight", "type": "float"},
    ]
    assert item.properties["table:row_count"] == 171
    assert "proj:bbox" not in item.properties
    assert item.geometry == {
        "coordinates": (
            (
                (29.53125, 9.795677582829734),
                (29.53125, 10.487811882056686),
                (28.828125, 10.487811882056686),
                (28.828125, 9.795677582829734),
                (29.53125, 9.795677582829734),
            ),
        ),
        "type": "Polygon",
    }
    assert item.bbox == [28.828125, 9.795677582829734, 29.53125, 10.487811882056686]


tests/test_module.py
import unittest

import stactools.msbuildings


class TestModule(unittest.TestCase):
    def test_version(self) -> None:
        self.assertIsNotNone(stactools.msbuildings.__version__)


tests/test_commands.py
import os.path
from tempfile import TemporaryDirectory
from typing import Callable, List

import pystac
from click import Command, Group
from stactools.testing.cli_test import CliTestCase

from stactools.bingbuildings.commands import create_bingbuildings_command


class CommandsTest(CliTestCase):
    def create_subcommand_functions(self) -> List[Callable[[Group], Command]]:
        return [create_bingbuildings_command]

    def test_create_collection(self) -> None:
        with TemporaryDirectory() as tmp_dir:
            # Run your custom create-collection command and validate

            # Example:
            destination = os.path.join(tmp_dir, "collection.json")

            result = self.run_command(f"bingbuildings create-collection {destination}")

            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))

            jsons = [p for p in os.listdir(tmp_dir) if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            collection = pystac.read_file(destination)
            self.assertEqual(collection.id, "my-collection-id")
            # self.assertEqual(item.other_attr...

            collection.validate()

    def test_create_item(self) -> None:
        with TemporaryDirectory() as tmp_dir:
            # Run your custom create-item command and validate

            # Example:
            infile = "/path/to/asset.tif"
            destination = os.path.join(tmp_dir, "item.json")
            result = self.run_command(
                f"bingbuildings create-item {infile} {destination}"
            )
            self.assertEqual(result.exit_code, 0, msg="\n{}".format(result.output))

            jsons = [p for p in os.listdir(tmp_dir) if p.endswith(".json")]
            self.assertEqual(len(jsons), 1)

            item = pystac.read_file(destination)
            self.assertEqual(item.id, "my-item-id")
            # self.assertEqual(item.other_attr...

            item.validate()


src/stactools/msbuildings/commands.py
from __future__ import annotations

import json
import logging
import pathlib

import click
from click import Command, Group

from stactools.msbuildings import stac

logger = logging.getLogger(__name__)


def create_msbuildings_command(cli: Group) -> Command:
    """Creates the stactools-msbuildings command line utility."""

    @cli.group(
        "msbuildings",
        short_help=("Commands for working with stactools-msbuildings"),
    )
    def msbuildings() -> None:
        pass

    @msbuildings.command(
        "create-collection",
        short_help="Creates a STAC collection",
    )
    @click.argument("destination")
    @click.option("--description", default=None, help="Collection description")
    @click.option(
        "--extra-field",
        default=None,
        help="Key-value pairs to include in extra-fields",
        multiple=True,
    )
    def create_collection_command(
        destination: str, description: str | None = None, extra_field: str | None = None
    ) -> None:
        """Creates a STAC Collection

        Args:
            destination (str): An HREF for the Collection JSON
        """
        extra_fields_d = dict(k.split("=") for k in extra_field)  # type: ignore

        collection = stac.create_collection(
            description=description, extra_fields=extra_fields_d
        )

        # collection.set_self_href(destination)
        pathlib.Path(destination).write_text(json.dumps(collection.to_dict(), indent=2))

    @msbuildings.command("create-item", short_help="Create a STAC item")
    @click.argument("source")
    @click.argument("destination")
    def create_item_command(source: str, destination: str) -> None:
        """Creates a STAC Item

        Args:
            source (str): HREF of the Asset associated with the Item
            destination (str): An HREF for the STAC Collection
        """
        item = stac.create_item(source)

        item.save_object(dest_href=destination)

        return None

    return msbuildings


src/stactools/msbuildings/geoparquet.py
from __future__ import annotations

import jsonschema
import requests
import fsspec
import geopandas
import pyarrow.fs
import pyproj
import json
from typing import Any

storage_options = dict(account_name="bingmlbuildings", credential="")
SCHEMA = "https://raw.githubusercontent.com/opengeospatial/geoparquet/v1.0.0-beta.1/format-specs/schema.json"


def add_geo_metadata(
    table: pyarrow.Table,
    geometry_types: list[str] | None = None,
    geometry_name: str = "geometry",
    geoparquet_version: str = "1.0.0-beta.1",
    schema_url: str | None = SCHEMA,
):
    if b"geo" in table.schema.metadata:
        # Already present
        return table

    geometry_types = geometry_types or []
    bbox = list(geopandas.array.from_wkb(table[geometry_name].to_numpy()).total_bounds)

    geo_metadata = {
        "version": geoparquet_version,
        "primary_column": geometry_name,
        "columns": {
            geometry_name: {
                "encoding": "WKB",
                "crs": json.loads(pyproj.CRS("WGS 84").to_json()),
                "geometry_types": geometry_types,
                "bbox": bbox,
            }
        },
    }
    metadata = {
        **table.schema.metadata,
        b"geo": json.dumps(geo_metadata).encode(),
    }

    if schema_url:
        r = requests.get(schema_url)
        r.raise_for_status()
        d = r.json()

        jsonschema.validate(geo_metadata, d)

    metadata = {
        **table.schema.metadata,
        b"geo": json.dumps(geo_metadata).encode(),
    }
    new_table = table.replace_schema_metadata(metadata)
    return new_table


def update_many(
    protocol: str,
    prefix: str,
    geometry_types: list[str],
    geometry_name: str = "geometry",
    geoparquet_version: str = "1.0.0-beta.1",
    storage_options: dict[str, Any] | None = None,
    schema_url: str | None = SCHEMA,
) -> str:
    """
    Add geoparquet metadata to all parquet files in a directory.
    """
    storage_options = storage_options or {}
    fs = fsspec.filesystem(protocol, **storage_options)
    arrow_fs = pyarrow.fs.PyFileSystem(pyarrow.fs.FSSpecHandler(fs))

    for path in fs.find(prefix):
        table = pyarrow.parquet.read_table(path, filesystem=fs)
        new_table = add_geo_metadata(
            table,
            geometry_types=geometry_types,
            geometry_name=geometry_name,
            geoparquet_version=geoparquet_version,
            schema_url=schema_url,
        )
        # TODO: robust error handling here
        pyarrow.parquet.write_table(new_table, path, filesystem=arrow_fs)
    return prefix


src/stactools/msbuildings/stac.py
from __future__ import annotations

import dataclasses
import functools
import importlib.resources
import json
import logging
from datetime import datetime, timezone
from typing import Any

import dateutil.parser
import mercantile
import shapely.geometry
import stac_table
from pystac import (
    Asset,
    CatalogType,
    Collection,
    Extent,
    Item,
    Link,
    MediaType,
    Provider,
    ProviderRole,
    RelType,
    SpatialExtent,
    TemporalExtent,
)
from pystac.extensions.item_assets import ItemAssetsExtension

logger = logging.getLogger(__name__)

ASSET_TITLE = "Building Footprints"
ASSET_DESCRIPTION = "Parquet dataset with the building footprints for this region."
GEOMETRY_DESCRIPTION = "Building footprint polygons"
# TODO: generalize to other storage
ASSET_EXTRA_FIELDS = {
    "table:storage_options": {"account_name": "bingmlbuildings"},
    "table:columns": [
        {
            "name": "geometry",
            "type": "byte_array",
            "description": GEOMETRY_DESCRIPTION,
        }
    ],
}


@functools.lru_cache
def get_data() -> dict[str, Any]:
    """
    Lookup metadata from a file provided by the footprints team.

    See `scripts/make_data.py`.
    """
    result = json.load(
        importlib.resources.open_text("stactools.msbuildings", "data.json")
    )
    assert isinstance(result, dict)
    return result


def create_collection(
    description: str | None = None, extra_fields: dict[str, Any] | None = None
) -> Collection:
    """Create a STAC Collection

    This function includes logic to extract all relevant metadata from
    an asset describing the STAC collection and/or metadata coded into an
    accompanying constants.py file.

    See `Collection<https://pystac.readthedocs.io/en/latest/api.html#collection>`_.

    Returns:
        Collection: STAC Collection object
    """
    providers = [
        Provider(
            name="Microsoft",
            roles=[ProviderRole.PRODUCER, ProviderRole.PROCESSOR, ProviderRole.HOST],
            url="https://planetarycomputer.microsoft.com",
        )
    ]

    start_datetime = datetime(2014, 1, 1, tzinfo=timezone.utc)

    extent = Extent(
        SpatialExtent([[-180.0, 90.0, 180.0, -90.0]]),
        TemporalExtent([[start_datetime, None]]),
    )

    links = [
        Link(
            rel=RelType.LICENSE,
            target="https://github.com/microsoft/GlobalMLBuildingFootprints/blob/main/LICENSE",
            media_type="text/html",
            title="ODbL-1.0 License",
        )
    ]
    keywords = [
        "Bing Maps",
        "Buildings",
        "geoparquet",
        "Microsoft",
        "Footprint",
    ]

    if description is None:
        description = (
            "Machine-learning detected building footprints. The underlying "
            "imagery is from Bing Maps and includes imagery from Maxar and Airbus."
        )

    collection = Collection(
        id="ms-buildings",
        title="Microsoft Building Footprints",
        description=description,  # noqa: E502
        license="ODbL-1.0",
        providers=providers,
        extent=extent,
        catalog_type=CatalogType.RELATIVE_PUBLISHED,
    )
    collection.links = links
    collection.keywords = keywords
    collection.add_asset(
        "thumbnail",
        Asset(
            "https://ai4edatasetspublicassets.blob.core.windows.net/assets/pc_thumbnails/msbuildings-thumbnail.png",  # noqa: E501
            title="Thumbnail",
            media_type=MediaType.PNG,
        ),
    )

    ItemAssetsExtension.add_to(collection)
    collection.extra_fields["item_assets"] = {
        "data": {
            "type": stac_table.PARQUET_MEDIA_TYPE,
            "title": ASSET_TITLE,
            "roles": ["data"],
            "description": ASSET_DESCRIPTION,
            **ASSET_EXTRA_FIELDS,
        }
    }

    if extra_fields:
        collection.extra_fields.update(extra_fields)

    return collection


@dataclasses.dataclass
class PathParts:
    """
    Parse a path like

    footprints/delta/2023-04-25/ml-buildings.parquet/RegionName=Abyei/quadkey=122321003/

    into its component parts.
    """

    path: str

    def __post_init__(self) -> None:
        split = self.path.rstrip("/").split("/")
        region_part = split[-2]
        datetime_part = split[-4]
        quadkey_part = split[-1]

        self.region = region_part.split("=", 1)[-1]
        self.datetime = dateutil.parser.parse(datetime_part).astimezone(timezone.utc)
        self.quadkey = int(quadkey_part.split("=")[1])


def create_item(
    asset_href: str,
    storage_options: dict[str, Any] | None = None,
    asset_extra_fields: dict[str, Any] | None = None,
) -> Item:
    """Create a STAC Item

    For

    Args:
        asset_href (str): The HREF pointing to an asset associated with the item

    Returns:
        Item: STAC Item object
    """
    parts = PathParts(asset_href)

    data = get_data().get(parts.region, {})

    start_datetime = data.get("MinCaptureDate")
    end_datetime = data.get("MaxCaptureDate")

    has_data = bool(data)
    properties = {
        "title": "Building footprints",
        "description": "Parquet dataset with the building footprints",
        "msbuildings:region": parts.region,
        "msbuildings:quadkey": parts.quadkey,
        "msbuildings:processing-date": parts.datetime.date().isoformat(),
    }

    tile = mercantile.quadkey_to_tile(str(parts.quadkey))
    bbox = mercantile.bounds(tile)
    shape = shapely.geometry.box(*bbox)

    geometry = shapely.geometry.mapping(shape)
    bbox = list(shape.bounds)

    if has_data:
        datetime = None
        properties.update(
            {"start_datetime": start_datetime, "end_datetime": end_datetime}
        )
    else:
        datetime = parts.datetime

    template = Item(
        id="_".join(
            [parts.region, str(parts.quadkey), parts.datetime.date().isoformat()]
        ),
        properties=properties,
        geometry=geometry,
        bbox=bbox,
        datetime=datetime,
        stac_extensions=[],
    )
    item = stac_table.generate(
        asset_href,
        template,
        storage_options=storage_options,
        asset_extra_fields=asset_extra_fields,
        infer_bbox=not has_data,
        infer_geometry=False,
        proj=False,
        count_rows=not has_data,
    )
    assert isinstance(item, Item)

    if has_data:
        item.properties["table:row_count"] = data["Count"]
    else:
        item.properties.pop("proj:bbox")

    item.properties["table:columns"][0]["description"] = GEOMETRY_DESCRIPTION

    # TODO: make configurable upstream
    item.assets["data"].title = ASSET_TITLE
    item.assets["data"].description = ASSET_DESCRIPTION

    return item
